sockeye/lexical_constraints.py

```diff
@@ -25,6 +25,104 @@
+
+    def consume(self, word_id: int) -> 'AvoidPhrase':
+        """
+        Consumes a word, and updates the tracking based on it. If the word is the next words in the phrase,
+        update last_consumed, otherwise, reset.
+
+        :param word_id: The word that was just generated.
+        """
+        if self.phrase[self.last_consumed + 1] == word_id:
+            return AvoidPhrase(self.phrase, self.last_consumed + 1)
+        else:
+            self.last_consumed = -1
+            return self
+
+    def avoid(self) -> int:
+        """
+        Returns an integer ID that should be avoided.
+
+        :return: An integer representing a word that must not be generated next by this hypothesis.
+        """
+        if self.last_consumed == len(self.phrase) - 2:
+            return self.phrase[-1]
+        else:
+            return 0
+
+
+class AvoidBatch:
+    """
+    Represents a set of phrasal constraints for all items in the batch.
+    For each hypotheses, there is a list of AvoidPhrase objects.
+
+    The offset is used to return actual positions in the one-dimensionally-resized array that
+    get set to infinity.
+
+    :param avoid_list: The list of lists (one for each item in the batch).
+    :param beam_size: The beam size.
+    :param max_id: The maximum ID in the target vocabulary.
+    """
+    def __init__(self,
+                 avoid_list: List[RawConstraintList],
+                 beam_size: int,
+                 max_id: int) -> None:
+        self.avoid_list = []  # type: List[List[AvoidPhrase]]
+        for raw_phrases in avoid_list:
+            self.avoid_list += [[AvoidPhrase(phrase) for phrase in raw_phrases] * beam_size]
+        self.offset = max_id
+
+    def reorder(self, indices: mx.nd.array) -> None:
+        """
+        Reorders the avoid list according to the selected row indices.
+        This can produce duplicates, but this is fixed in consume().
+
+        :param indices: An mx.nd.array containing indices of hypotheses to select.
+        """
+        self.avoid_list = [self.avoid_list[x] for x in indices.asnumpy()]
+
+    def consume(self, word_ids: mx.nd.array) -> None:
+        """
+        Takes an array representing the set of target word indices chosen for each hypothesis in the batch,
+        and updates each of the negative constraints.
+
+        :param word_ids: The set of word IDs.
+        """
+        word_ids = word_ids.asnumpy().tolist()
+        for i, word_id in enumerate(word_ids):
+            self.avoid_list[i] = [item.consume(word_ids[i]) for item in self.avoid_list[i]]
+
+    def avoid(self) -> List[int]:
+        """
+        Returns a list of indices in a 1-dimensional reshaping of the scores array that should be set to null.
+
+        :return: A list of indices.
+        """
+        to_avoid = []
+        for i, items in enumerate(self.avoid_list):
+            for phrase in items:
+                word_id = phrase.avoid()
+                if word_id:
```

Comment:
Not strictly necessary, but I would be explicit here and put `if word_id > 0:', as it cannot be None.